src/coreclr/vm/comsynchronizable.cpp

```diff
@@ -795,41 +801,49 @@ void ThreadNative::InformThreadNameChange(Thread* pThread, LPCWSTR name, INT32 l
 #endif // DEBUGGING_SUPPORTED
 }
 
-FCIMPL1(FC_BOOL_RET, ThreadNative::IsThreadpoolThread, ThreadBaseObject* thread)
+// Get whether or not this is a threadpool thread.
+extern "C" BOOL QCALLTYPE ThreadNative_GetIsThreadPoolThread(QCall::ThreadHandle thread)
 {
-    FCALL_CONTRACT;
+    CONTRACTL
+    {
+        QCALL_CHECK;
+        PRECONDITION(thread != NULL);
+    }
+    CONTRACTL_END;
 
-    if (thread==NULL)
-        FCThrowRes(kNullReferenceException, W("NullReference_This"));
+    BOOL res = FALSE;
 
-    Thread *pThread = thread->GetInternal();
+    BEGIN_QCALL;
 
-    if (pThread == NULL)
-        FCThrowRes(kThreadStateException, W("ThreadState_Dead_State"));
+    if (ThreadIsDead(thread))
+        COMPlusThrow(kThreadStateException, W("ThreadState_Dead_State"));
 
-    BOOL ret = pThread->IsThreadPoolThread();
+    res = thread->IsThreadPoolThread();
 
-    FC_GC_POLL_RET();
+    END_QCALL;
 
-    FC_RETURN_BOOL(ret);
+    return res;
 }
-FCIMPLEND
 
-FCIMPL1(void, ThreadNative::SetIsThreadpoolThread, ThreadBaseObject* thread)
+// Set thread as a threadpool thread.
+extern "C" void QCALLTYPE ThreadNative_SetIsThreadPoolThread(QCall::ThreadHandle thread)
 {
-    FCALL_CONTRACT;
+    CONTRACTL
+    {
+        QCALL_CHECK;
+        PRECONDITION(thread != NULL);
+    }
+    CONTRACTL_END;
 
-    if (thread == NULL)
-        FCThrowResVoid(kNullReferenceException, W("NullReference_This"));
+    BEGIN_QCALL;
 
-    Thread *pThread = thread->GetInternal();
+    if (ThreadIsDead(thread))
```

Comment:
This can be an assert (potentially on the managed side). It is internal setter that is only called from our own code, so we should not be calling it on dead threads.